### DISCUSSION TITLE
ValidPath without const assertion?

### DISCUSSION THREAD

**Discussion by asgerhallas** [millsp/ts-toolbelt#178] [Wiki] [answered]
## 🤔 Question

I'm new to this library, and am still trying to get my head around it. It's very impressive! :)

In this sample https://pirix-gh.github.io/ts-toolbelt/modules/_object_pathvalid_.html there's the following code: 

    const test0 = getAt(state,          ['case', 'name'] as const) // {c: number}

Why is const needed? And is there a way to make it work without the const assertion?

Hope you can help.

**Comment by millsp** [comment]
Hi there,

Without a const assertion, the type of `['case', 'name']` would be narrowed to `string[]` that's all. Depending on your type safety needs, you would choose `as const` or not.

**Comment by asgerhallas** [comment]
I somehow missed your answer, sorry! What I meant was that, I'd like the type safety to be enforced by the function signature, so the caller does not have to add `as const`, but still only be able to pass expected strings in the array. It's easy for a caller to forget `as const` and thus we will not catch errors now or when the allowed string values change in the future.

**Comment by millsp** [comment]
It's a TypeScript limitation, I know how inconvenient this is. To work around that, the only solution is to take rest parameters instead. Or you could have this Maybe this can help you:

- https://github.com/millsp/ts-toolbelt/issues/154
- https://github.com/microsoft/TypeScript/issues/30680

**Comment by asgerhallas** [comment]
Thanks a lot for your swift reply :) 

I'll look into those links right away!

**Comment by millsp** [comment]
I posted a solution to this problem here in https://github.com/microsoft/TypeScript/issues/30680. It will land soon in ts-toolbelt.

**Comment by asgerhallas** [comment]
This is amazing, thank you for doing this! I look forward to it landing is ts-toolbelt too.

**Comment by millsp** [comment, marked as answer]
This has been shipped in the latest release https://millsp.github.io/ts-toolbelt/modules/function_narrow.html

**Reply by schtauffen** [reply]
Is it possible to have the completion work on the path AND have the Narrow function?
```ts
declare function getAt<
  T extends object,
  P extends L.List<A.Key>
>(lens: Narrow<P>, target: T): O.Path<T, P>;

getAt(["foo"], { foo: 3 }); // ✓ number
getAt(["bar"], { foo: 3 }); // ✓ never

declare function get<
  T extends object,
  P extends O.Paths<T>
>(lens: Narrow<P>, target: T): O.Path<T, P>;

get(["bar"], { foo: 3 }); // ✓ error
get(["foo"], { foo: 3 }); // X { foo: number } :(
```

**Reply by schtauffen** [reply]
I feel like a tried a million combinations and hit on this almost immediately after posting:
```ts
declare function getB<
  T extends object,
  P extends L.List<A.Key>
>(lens: A.Cast<Narrow<P>, O.Paths<T>>, target: T): O.Path<T, P>;

getB(["bar"], { foo: 3 }); // ✓ error
getB(["foo"], { foo: 3 }); // ✓ number
```
